modules/logging/ui/src/sap/commerce/toolset/logging/actionSystem/CxCreateCustomTemplateStateAction.kt

```diff
@@ -0,0 +1,174 @@
+        val uniqueLoggers = loggers
+            .fold(linkedMapOf<String, CxLoggerPresentation>()) { mergedLoggers, log ->
+                mergedLoggers[log.name] = log
+                mergedLoggers
+            }
+            .values
+            .toList()
+
+        val customTemplate = customLogTemplateService.createTemplateFromLoggers(templateName, uniqueLoggers).mutable()
+        val dialogContext = LogTemplateDialogContext(
+            project = project,
+            mutable = customTemplate,
+            title = dialogTitle,
+            duplicatedSourceTemplates = loggers.size != uniqueLoggers.size,
+        )
+        val dialog = when (selectedNodes.first()) {
+            is CxRemoteLogStateNode -> CxCustomLogTemplateDialog(dialogContext)
+            is CxBundledLogTemplateItemNode -> CxCustomLogTemplateDialog(dialogContext)
+            is CxCustomLogTemplateItemNode -> CxAdvancedCustomLogTemplateDialog(dialogContext)
+            else -> null
+        } ?: return
+
+        if (dialog.showAndGet()) {
+            customLogTemplateService.addTemplate(customTemplate.immutable())
+
+            if (dialogContext.removeSourceTemplates.get()) {
+                val customTemplateUids = selectedNodes
+                    .mapNotNull { it.asSafely<CxCustomLogTemplateItemNode>() }
+                    .map { it.uuid }
+                customLogTemplateService.deleteTemplates(customTemplateUids)
+            }
+        }
+    }
+
+    private fun getDialogTitle(selectedNodes: List<Any>) = when (selectedNodes.first()) {
+        is CxRemoteLogStateNode -> "Create a Log Template"
+        is CxBundledLogTemplateItemNode -> if (selectedNodes.size > 1) "Merge Templates" else "Clone Template"
+        is CxCustomLogTemplateItemNode -> if (selectedNodes.size > 1) "Merge Templates" else "Clone Template"
+        else -> "Create a Log Template"
+    }
+
+    private fun templateNameFromConnection(project: Project, connectionName: String): String {
+        val templateName = "Remote '$connectionName' | template"
+        val count = CxCustomLogTemplatesSettings.getInstance(project).templates.count { it.name.startsWith(templateName) }
+
+        return if (count >= 1) "$templateName ($count)"
+        else templateName
+    }
+
+    private fun loggers(project: Project, selectedNodes: List<Any>) = selectedNodes
+        .mapNotNull {
+            when (it) {
+                is CxRemoteLogStateNode -> CxRemoteLogStateService.getInstance(project).state(it.connection.uuid).get()?.values ?: emptyList()
+                is CxBundledLogTemplateItemNode -> it.loggers
+                is CxCustomLogTemplateItemNode -> it.loggers
+                else -> null
+            }
+        }
+        .flatten()
+
+    private fun templateName(project: Project, selectedNodes: List<Any>) = when (val firstNode = selectedNodes.first()) {
+        is CxRemoteLogStateNode -> templateNameFromConnection(project, firstNode.connection.shortenConnectionName)
```

Comment:
it is better to place methods in the order of usage, `templateNameFromConnection` can be placed below this one